docs/design/mono/embedding-api-deprecation.md

```diff
@@ -0,0 +1,68 @@
+MONO_API MONO_RT_EXTERNAL_ONLY
+MonoAssembly* mono_assembly_load_full (MonoAssemblyName *aname, const char *basedir, MonoImageOpenStatus *status, mono_bool refonly);
+MONO_API MONO_RT_EXTERNAL_ONLY
+MonoAssembly* mono_assembly_load_from (MonoImage *image, const char *fname, MonoImageOpenStatus *status);
+MONO_API MONO_RT_EXTERNAL_ONLY
+MonoAssembly* mono_assembly_load_from_full (MonoImage *image, const char *fname, MonoImageOpenStatus *status, mono_bool refonly);
+MONO_API MONO_RT_EXTERNAL_ONLY
+MonoAssembly* mono_assembly_load_with_partial_name (const char *name, MonoImageOpenStatus *status);
+```
+
+Note that *none* of those are even what we use internally! And none of them offer AssemblyLoadContext support, so the list isn't even complete.
+
+# Current state in .NET 5
+
+With the migration to dotnet/runtime and .NET 5 changes, many of the embedding APIs are no longer relevant. Currently the behavior with inapplicable APIs is inconsistent, and they will either do nothing, error, or silently ignore arguments.
+
+There have also been some small breaks in the behavior of existing APIs, in particular some loader-related changes documented [here](https://www.mono-project.com/news/2020/08/24/native-loader-net5/).
+
+Support for the `coreclrhost.h` APIs already exists in Mono, as the default `dotnet` host relies on them to function. Those are not defined in a shipping header, and instead users are expected to copy over the definitions and trust us to export the symbols. The relevant file is [mini/main-core.c](https://github.com/dotnet/runtime/blob/master/src/mono/mono/mini/main-core.c).
+
+We have also been forced to add a handful of .NET 5-specific API extensions, mostly to offer the Xamarin products AssemblyLoadContext-aware versions of preexisting APIs. We have defined those in the `mono-private-unstable.h` headers, but this is intended as a temporary measure to unblock some .NET 5 migrations.
+
+# Moving forward
+
+For .NET 6, we have an opportunity to clean up the existing headers and define a clear policy around ABI/API stability and deprecation.
+
+## ABI stability
+
+ABI stability is only guaranteed across major version releases. .NET 6 will not receive a servicing update that breaks ABI compatibility, but you cannot link a .NET 5 app against the .NET 6 release, full stop.
+
+## API stability
+
+API stability is more complicated. There are two distinct cases here: .NET 6 and the behavior moving forward.
+
+With the .NET 6 release specifically, we have an opportunity to do some initial API cleanup work because relatively few desktop users have migrated over to .NET 5 Mono. There is some low-hanging fruit to potentially clean up immediately, before we have significant migration: irrelevant functionality like `mono_domain_set` or `refonly` parameters, legacy compatibility hacks like `MonoLookupPInvokeStatus`, and subtly-incorrect behavior like fixed-size GC handles. 
```

Comment:
I'm not sure that "most embedding users are on net4x, and therefore we can set the bar for migration higher" is a compelling argument.  To me it sounds like net6 would have to be solely focused deprecation and offering an _optional_ migration path, but not outright removing symbols.

Although in some cases the deprecated symbols might have semantic changes (e.g. that `refonly` parameter when it's not-FALSE can't possibly have the same behavior as in net4x).

Reply:
Hmm. Maybe I'm just trying to justify a desire to immediately get rid of stuff that I think should have been removed years ago like `MonoLookupPInvokeStatus`, and that's not the right balance here. I dunno though, I think that with .NET Core there's precedent for "rip out the stuff that makes no sense under the new design" and/or should been gone years ago, and I'm not actually sure we're making life that much easier for embedders if we continue to allow calls to things like `mono_domain_set`...

Reply:
> I'm not actually sure we're making life that much easier for embedders if we continue to allow calls to things like `mono_domain_set`

That's not really a typical example, though.  There's a clear easy good replacement for `mono_domain_set` - a no-op.

Some function that returns a value that makes sense in some circumstances but not others may be less obvious.  Take `mono_lookup_pinvoke_call`, for example, if I have a ton of code to handle the "exception class, exception arg" out-arg case, is it obvious how to rewrite it in terms of something that uses `MonoError` (presumably our replacement API)?

(digression: I assume you don't literally mean `MonoLookupPInvokeStatus` - it isn't a public API - but `mono_lookup_pinvoke_call` which is the ugly API with the exception name out parameter that `MonoLookupPInvokeStatus` is meant to paper over)

Also what is the upgrade path from net4x to net6 like now for an embedder if we remove functions: you have to work around everything we removed before you can even try your code to see if it still works with System.Private.CoreLib / the new BCL / new managed API surface etc.